migrations/2025-03-06-060640_semver_ord/up.sql

```diff
@@ -0,0 +1,109 @@
+        -- the `prerelease` array.
+        prerelease_parts := string_to_array(match_result[4], '.');
+
+        foreach prerelease_part in array prerelease_parts[1:max_prerelease_parts + 1]
+        loop
+            -- Parse parts as numbers if they consist of only digits.
+            if regexp_like(prerelease_part, '^\d+$') then
+                -- In JSONB a number has higher precedence than a string but in
+                -- semver it is the other way around, so we use true/false to
+                -- work around this.
+                prerelease := jsonb_set(prerelease, array[i::text], to_jsonb(false));
+                prerelease := jsonb_set(prerelease, array[(i + 1)::text], to_jsonb(prerelease_part::numeric));
+            else
+                prerelease := jsonb_set(prerelease, array[i::text], to_jsonb(true));
+                prerelease := jsonb_set(prerelease, array[(i + 1)::text], to_jsonb(prerelease_part));
+            end if;
+
+            -- Exit the loop if we have reached the maximum number of parts.
+            i := i + 2;
+            exit when i >= max_prerelease_parts * 2;
+        end loop;
+
+        prerelease := jsonb_set(prerelease, array[(max_prerelease_parts * 2)::text], to_jsonb(array_to_string(prerelease_parts[max_prerelease_parts + 1:], '.')));
+    end if;
+
+    -- Return an array with the major, minor, patch, and prerelease parts.
+    return json_build_array(
+        match_result[1]::numeric,
+        match_result[2]::numeric,
+        match_result[3]::numeric,
+        prerelease
+    );
+end;
+$$ language plpgsql immutable;
+
+comment on function semver_ord is 'Converts a semver string into a JSONB array for version comparison purposes. The array has the following format: [major, minor, patch, prerelease] and when used for sorting follow the precedence rules defined in the semver specification (https://semver.org/#spec-item-11).';
+
+
+-- Add corresponding column to the `versions` table.
+
+alter table versions
+    add semver_ord jsonb;
+
+comment on column versions.semver_ord is 'JSONB representation of the version number for sorting purposes.';
+
+
+-- Create a trigger to set the `semver_ord` column when inserting a new version.
+-- Ideally, we would use a generated column for this, but introducing such a
+-- column would require a full table rewrite, which is not feasible for large
+-- tables.
+
+create or replace function set_semver_ord() returns trigger as $$
+begin
+    new.semver_ord := semver_ord(new.num);
+    return new;
+end
+$$ language plpgsql;
+
+create or replace trigger trigger_set_semver_ord
+    before insert on versions
+    for each row
+    execute procedure set_semver_ord();
+
+
+-- Populate the `semver_ord` column for existing versions.
+-- This query should be run manually in small batches to avoid locking the
+-- table for too long.
+
+-- with versions_to_update as (
+--     select id, num
+--     from versions
+--     where semver_ord = 'null'::jsonb
```

Comment:
I just noticed that I didn't update this correctly. this line should now be `where semver_ord is null`.